### PR TITLE
Remove the usage of main mediator in MainActivity

### PR DESCRIPTION
**Pull request by pharspher**

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.java [hunk 1]
```diff
@@ -778,30 +776,18 @@ public void onNotified(@NonNull Fragment from, @NonNull TYPE type, @Nullable Obj
...
+                this.screenNavigator.popUrlScreen();
                 break;
             case SHOW_TAB_TRAY:
-                if ((payload != null) && (payload instanceof String)) {
-                    TabTray.show(getSupportFragmentManager(), (String) payload);
-                }
+                TabTray.show(getSupportFragmentManager());
                 break;
             case REFRESH_TOP_SITE:
-                HomeFragment homeFragment = this.mainMediator.getTopHomeFragment();
```
Comment (pharspher) [inline app/src/main/java/org/mozilla/focus/activity/MainActivity.java]: Note: revisit this before merge

File: app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java [hunk 2]
```diff
@@ -21,12 +32,31 @@
  * 1. Move more methods from MainMediator to ScreenNavigator
  * 2. Finally remove MainMediator from MainActivity
  */
-public class ScreenNavigator {
+public class ScreenNavigator implements LifecycleObserver {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java]: f you use Java 8 Language, then observe events with DefaultLifecycleObserver. To include it you should add "android.arch.lifecycle:common-java8:<version>" to your build.gradle file. 

If you use Java 7 Language, Lifecycle events are observed using annotations. Once Java 8 Language becomes mainstream on Android, annotations will be deprecated, so between DefaultLifecycleObserver and annotations, you must always prefer DefaultLifecycleObserver.

File: app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java [hunk 3]
```diff
@@ -40,9 +70,25 @@ public static ScreenNavigator get(Context context) {
         }
     }
 
-    ScreenNavigator(MainActivity activity, MainMediator mediator) {
-        this.mainMediator = mediator;
-        this.browserMediator = new BrowserMediator(activity, mainMediator);
+    ScreenNavigator(@Nullable MainActivity activity) {
+        if (activity == null) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java]: Is it valid to create an empty ScreenNavigator with no mediator..? Maybe throwing Exception here?

File: app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java [hunk 4]
```diff
@@ -96,8 +152,71 @@ public void addHomeScreen(boolean animate) {
     public void popToHomeScreen(boolean animate) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java]: consider sharing code in addHomeScreen() and popToHomeScreen()

File: app/src/main/java/org/mozilla/focus/activity/MainMediator.java [hunk 5]
```diff
@@ -98,90 +92,10 @@ boolean shouldFinish() {
...
-
-            @Override
-            public void onAnimationRepeat(Animation animation) {
-            }
-        });
-        view.startAnimation(fadeOut);
-    }
-
-    void onFragmentStarted(@NonNull String tag) {
-        if (UrlInputFragment.FRAGMENT_TAG.equals(tag)) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/activity/MainMediator.java]: I'm a bit lost here. Why don't we just do toggleFakeUrlInput() in UrlInputFragment's onStart?

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.java [hunk 6]
```diff
@@ -151,9 +149,9 @@ protected void onCreate(Bundle savedInstanceState) {
                 pendingUrl = intent.getStringExtra(RocketMessagingService.PUSH_OPEN_URL);
             } else {
                 if (Settings.getInstance(this).shouldShowFirstrun()) {
-                    this.mainMediator.showFirstRun();
+                    this.screenNavigator.addFirstRunScreen();
                 } else {
-                    this.mainMediator.showHomeScreen();
+                    this.screenNavigator.popToHomeScreen(false);
```
[the remark this answers is not in the document]
Comment (pharspher) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.java]: there's nothing in the back stack at the phase, so there's nothing gonna be popped. BTW, BrowserFragment is embedded in MainActivity, so it will never be any transaction that related to BrowserFragment.

File: app/src/main/java/org/mozilla/focus/activity/MainMediator.java [hunk 7]
```diff
@@ -251,7 +165,7 @@ private void toggleFakeUrlInput(boolean visible) {
     /**
      * get HomeFragment if it's Top Fragment
      */
-    HomeFragment getTopHomeFragment() {
+    private HomeFragment getTopHomeFragment() {
```
[the remark this answers is not in the document]
Comment (pharspher) [reply app/src/main/java/org/mozilla/focus/activity/MainMediator.java]: removed in later commit

File: app/src/main/java/org/mozilla/focus/activity/MainMediator.java [hunk 8]
```diff
@@ -98,90 +92,10 @@ boolean shouldFinish() {
...
-
-    void onFragmentStopped(@NonNull String tag) {
-        if (UrlInputFragment.FRAGMENT_TAG.equals(tag)) {
-            toggleFakeUrlInput(true);
-        }
+        return TYPE_ROOT.equals(getEntryType(lastEntry));
     }
 
-    Fragment getTopFragment() {
+    private Fragment getTopFragment() {
```
[the remark this answers is not in the document]
Comment (pharspher) [reply app/src/main/java/org/mozilla/focus/activity/MainMediator.java]: done

File: app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java (hunk 2 above)
Comment (pharspher) [reply app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java]: I'll create another PR for this

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mTwTm) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.java]: Oh I mean HomeFragment sorry my bad :P

File: app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java (hunk 3 above)
Comment (pharspher) [reply app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java]: On debug version, this will indeed lead to crash when the static factory of ScreenNavigator is called.
I allowed null here to make it easier to write a no-op version subclass.

Comment (pharspher) [reply app/src/main/java/org/mozilla/focus/activity/ScreenNavigator.java]: I'll see if i can improve this on another PR

File: app/src/main/java/org/mozilla/focus/activity/MainMediator.java (hunk 5 above)
Comment (pharspher) [reply app/src/main/java/org/mozilla/focus/activity/MainMediator.java]: I don't think UrlInputFragment should know when or how a specific UI element on home fragment would behave.

File: app/src/main/java/org/mozilla/focus/activity/MainMediator.java [hunk 9]
```diff
@@ -223,4 +207,75 @@ public void onBackStackChanged() {
...
+
+            logData("addNewEntry");
+            return key;
+        }
+
+        EntryData get(FragmentManager.BackStackEntry entry) {
+            return data.get(entry.getName());
+        }
+
+        void onBackStackChanged(FragmentManager manager) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/activity/MainMediator.java]: I don't get why we're splitting this into two methods (onBackStackChanged and purge) and why are we maintaining two copies of data. (For concurrency problems?)

File: app/src/main/java/org/mozilla/focus/navigation/TransactionHelper.java [hunk 10]
```diff
@@ -123,24 +169,22 @@ void toggleFakeUrlInput(boolean visible) {
...
+    private String getEntryTag(FragmentManager.BackStackEntry entry) {
         return entryDataSet.get(entry).tag;
     }
 
     private @EntryData.EntryType int getEntryType(FragmentManager.BackStackEntry entry) {
         return entryDataSet.get(entry).type;
     }
 
-    public void updateForegroundType(@EntryData.EntryType int type) {
+    private boolean isStateSaved() {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/navigation/TransactionHelper.java]: You're doing things that are not actually `Rename MainMedaitor to TransactionHelper`. Consider moving these to another commit.

Comment (pharspher) [reply app/src/main/java/org/mozilla/focus/navigation/TransactionHelper.java]: done

File: app/src/main/java/org/mozilla/focus/activity/MainMediator.java (hunk 9 above)
Comment (pharspher) [reply app/src/main/java/org/mozilla/focus/activity/MainMediator.java]: I'll move the call to purge() into onBackStackChanged()

File: app/src/main/java/org/mozilla/focus/navigation/TransactionHelper.java [hunk 11]
```diff
@@ -238,39 +237,32 @@ public void onBackStackChanged() {
...
             logData("addNewEntry");
             return key;
         }
 
         EntryData get(FragmentManager.BackStackEntry entry) {
-            return data.get(entry.getName());
+            return dataMap.get(entry.getName());
         }
 
         void onBackStackChanged(FragmentManager manager) {
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/navigation/TransactionHelper.java]: Maybe rename this to purge.
